equational_theories/MemoFinOp.lean

```diff
@@ -0,0 +1,67 @@
+import Lean
+
+/-!
+This file defines a macro `memoFinOp` that takes a function `f : Fin n → Fin n → Fin n`, at
+elaboration time calculates its table of values, encodes it into a single `Nat`, and returns a term
```

Comment:
Is the trick here that nat arithmetic "lookups" are faster than storing a flat array? Is kernel reduction speed the relevant performance metric here, or runtime evaluation?

Reply:
Yes, in the kernel there are no arrays, they become lists, and then lookup is likely quite slow.

I think kernel reduction is the right metric, as long as we use `by decide` and not `by native_decide`

Reply:
Wasn't the conclusion from Zulip that it's better to use a base which is a power of two, with shifts and bitwise and to extract the data?